lua/telescope/builtin/files.lua

```diff
@@ -289,134 +289,120 @@ local function prepare_match(entry, kind)
+    end,
+  })
+  table.insert(data, 1, ".." .. os_sep)
+  -- returns copy with properly set cwd for entry maker
+  return finders.new_table { results = data, entry_maker = opts.entry_maker { cwd = opts.path } }
+end
 
-          return rawget(t, rawget({ value = 1 }, k))
-        end
+local browse_folders = function(opts)
+  local data = {}
+  scan.scan_dir(opts.cwd, {
+    hidden = opts.hidden,
+    only_dirs = true,
+    on_insert = function(entry)
+      table.insert(data, entry .. os_sep)
+    end,
+  })
+  table.insert(data, 1, "." .. os_sep)
+  return finders.new_table { results = data, entry_maker = opts.entry_maker { cwd = opts.cwd } }
+end
 
-        return function(line)
-          local tbl = { line }
-          tbl.ordinal = Path:new(line):make_relative(opts.cwd)
-          return setmetatable(tbl, mt)
+local fb_finder = function(opts)
+  local cwd = vim.loop.cwd()
+  return setmetatable({
+    path = vim.F.if_nil(opts.path, cwd), -- current path for file browser
+    cwd = vim.F.if_nil(opts.cwd, cwd), -- nvim cwd
+    hidden = vim.F.if_nil(opts.hidden, false),
+    depth = vim.F.if_nil(opts.depth, 1), -- depth for file browser
+    files = vim.F.if_nil(opts.files, true), -- file or folders mode
+    -- ensure we forward make_entry opts adequately
+    entry_maker = vim.F.if_nil(opts.entry_maker, function(local_opts)
+      return make_entry.gen_from_fs(vim.tbl_extend("force", opts, local_opts))
+    end),
+    _file_finder = browse_files,
+    _folder_finder = browse_folders,
+    _cached_folder_finder = false,
+    _was_hidden = vim.F.if_nil(opts.hidden, false),
+    close = function() end,
+  }, {
+    __call = function(self, ...)
+      if self.files then
+        self._finder = self._file_finder(self)
+      else
+        local cwd_ = vim.loop.cwd()
+        if self._cached_folder_finder == false or cwd_ ~= self.cwd then
+          self._cached_folder_finder = self._folder_finder(self)
+          self.hidden = self._was_hidden
+          self.cwd = cwd_
         end
+        self._finder = self._cached_folder_finder
+      end
+      self._finder(...)
+    end,
+    __index = function(self, k)
+      if self._finder[k] ~= nil then
+        return self._finder[k]
+      else
+        error(string.format("%s not in finder", k))
       end
+    end,
+  })
+end
 
-      return finders.new_table { results = data, entry_maker = maker() }
-    end
+files.file_browser = function(opts)
+  opts = opts or {}
+
+  local cwd = vim.loop.cwd()
+  opts.depth = opts.depth or 1
+  opts.cwd = opts.cwd and vim.fn.expand(opts.cwd) or cwd
+  opts.files = vim.F.if_nil(opts.files, true)
 
   pickers.new(opts, {
-    prompt_title = "File Browser",
-    finder = opts.new_finder(opts.cwd),
+    prompt_title = opts.files and "File Browser" or "Folder Browser",
+    results_title = opts.files and Path:new(opts.cwd):make_relative(cwd) .. os_sep or "Results",
+    finder = fb_finder(opts),
     previewer = conf.file_previewer(opts),
```

Comment:
got the following error while trying to preview a root directory with no user access to read.

```lua
Error executing vim.schedule lua callback: .../pack/packer/start/plenary.nvim/lua/plen
ary/filetype.lua:103: bad argument #1 to 'ipairs' (table expected, got nil)
```

Reply:
😆 hit c-g until I got to my real home 😆